tests/app/questionnaire/test_value_source_resolver.py

```diff
@@ -0,0 +1,371 @@
+    **answer_source,
+    "selector": "years",
+}
+answer_source_list_item_selector_location = {
+    **answer_source,
+    "list_item_selector": {"source": "location", "id": "list_item_id"},
+}
+answer_source_list_item_selector_list_first_item = {
+    **answer_source,
+    "list_item_selector": {"source": "list", "id": "some-list", "id_selector": "first"},
+}
+
+metadata_source = {"source": "metadata", "identifier": "some-metadata"}
+
+list_source = {"source": "list", "identifier": "some-list"}
+list_source_id_selector_first = {**list_source, "id_selector": "first"}
+list_source_id_selector_primary_person = {
+    **list_source,
+    "id_selector": "primary_person",
+}
+list_source_id_selector_same_name_items = {
+    **list_source,
+    "id_selector": "same_name_items",
+}
+
+location_source = {"source": "location", "identifier": "list_item_id"}
+
+
+def get_list_items(num: int):
+    return [f"item-{i}" for i in range(1, num + 1)]
+
+
+def get_mock_schema():
+    schema = Mock(
+        QuestionnaireSchema(
+            {
+                "questionnaire_flow": {
+                    "type": "Linear",
+                    "options": {"summary": {"collapsible": False}},
+                }
+            }
+        )
+    )
+    return schema
+
+
+def get_value_source_resolver(
+    schema: QuestionnaireSchema = None,
+    answer_store: AnswerStore = AnswerStore(),
+    list_store: ListStore = ListStore(),
+    metadata: Optional[dict] = None,
+    location: Union[Location, RelationshipLocation] = Location(
+        section_id="test-section", block_id="test-block"
+    ),
+    list_item_id: Optional[str] = None,
+    routing_path_block_ids: Optional[list] = None,
+    use_default_answer=False,
+):
+    if not schema:
+        schema = get_mock_schema()
+        schema.answer_should_have_list_item_id = Mock(return_value=bool(list_item_id))
+
+    if not use_default_answer:
+        schema.get_default_answer = Mock(return_value=None)
+
+    return ValueSourceResolver(
+        answer_store=answer_store,
+        list_store=list_store,
+        metadata=metadata,
+        schema=schema,
+        location=location,
+        list_item_id=list_item_id,
+        routing_path_block_ids=routing_path_block_ids,
+        use_default_answer=use_default_answer,
+    )
+
+
+def test_answer_source():
+    value_source_resolver = get_value_source_resolver(
+        answer_store=AnswerStore([{"answer_id": "some-answer", "value": "Yes"}]),
+    )
+
+    assert value_source_resolver.resolve(answer_source) == "Yes"
+
+
+def test_answer_source_with_dict_answer_selector():
+    value_source_resolver = get_value_source_resolver(
+        answer_store=AnswerStore(
+            [
+                {
+                    "answer_id": "some-answer",
+                    "value": {"years": 1, "months": 10},
+                }
+            ]
+        ),
+    )
+
+    assert value_source_resolver.resolve(answer_source_dict_answer_selector) == 1
```

Comment:
That would create a lot of duplication which is what I was trying to avoid, but at the same time understand where you are coming from. Did you mean just the selectors, or even the parameterised lists as well?.

Reply:
Just the selectors. I think duplication is acceptable to make the tests readable.

Reply:
Have defined the selectors inline.